packages/zoltan2/src/algorithms/color/Zoltan2_AlgHybridGMB.hpp

```diff
@@ -0,0 +1,458 @@
+      
+      RCP<const map_t> mapOwned;
+      RCP<const map_t> mapWithCopies;
+      
+      std::vector<gno_t> finalGIDs;
+      std::vector<offset_t> finalOffset_vec;
+      std::vector<lno_t> finalAdjs_vec;      
+
+      lno_t nInterior = 0;
+      std::vector<lno_t> reorderToLocal;
+      for(size_t i = 0;  i< nVtx; i++) reorderToLocal.push_back(i);
+      
+      //Set up a typical local mapping here.
+      std::unordered_map<gno_t,lno_t> globalToLocal;
+      std::vector<gno_t> ownedPlusGhosts;
+      for (int i = 0; i < vtxIDs.size(); i++){
+        globalToLocal[vtxIDs[i]] = i;
+        ownedPlusGhosts.push_back(vtxIDs[i]);
+      }
+      int nghosts = 0;
+      for (int i = 0; i < adjs.size(); i++){
+        if(globalToLocal.count(adjs[i]) == 0){
+          //new unique ghost found
+          ownedPlusGhosts.push_back(adjs[i]);
+          globalToLocal[adjs[i]] = vtxIDs.size() + nghosts;
+          nghosts++;
+            
+        }
+      }
+        
+      finalAdjs_vec.resize(adjs.size()); 
+      for(size_t i = 0; i < finalAdjs_vec.size();i++){
+        finalAdjs_vec[i] = globalToLocal[adjs[i]];
+      }
+      for(int i = 0; i < offsets.size(); i++) finalOffset_vec.push_back(offsets[i]);
+      finalGIDs = ownedPlusGhosts;
+
+      Tpetra::global_size_t dummy = Teuchos::OrdinalTraits
+                                             <Tpetra::global_size_t>::invalid();
+      mapOwned = rcp(new map_t(dummy, vtxIDs, 0, comm));
+
+      dummy = Teuchos::OrdinalTraits <Tpetra::global_size_t>::invalid();
+      mapWithCopies = rcp(new map_t(dummy, 
+                                Teuchos::arrayViewFromVector(ownedPlusGhosts),
+                                0, comm)); 
+                                      
+      //create the FEMultiVector for the distributed communication.
+      //We also use the views from this datastructure as arguments to
+      //KokkosKernels coloring functions.
+      typedef Tpetra::Import<lno_t, gno_t> import_t;
+      Teuchos::RCP<import_t> importer = rcp(new import_t(mapOwned, 
+                                                            mapWithCopies));
+      Teuchos::RCP<femv_t> femv = rcp(new femv_t(mapOwned, 
+                                                    importer, 1, true));
+      timeReorder->stop();
+      //Get color array to fill
+      ArrayRCP<int> colors = solution->getColorsRCP();
+      for(size_t i=0; i<nVtx; i++){
+        colors[i] = 0;
+      } 
+      
+      //Create random numbers seeded on global IDs so that we don't
+      //need to communicate for consistency. These numbers determine
+      //which vertex gets recolored in the event of a conflict.
+      //taken directly from the Zoltan coloring implementation 
+      std::vector<int> rand(finalGIDs.size());
+      for(size_t i = 0; i < finalGIDs.size(); i++){
+        Zoltan_Srand((unsigned int) finalGIDs[i], NULL);
+        rand[i] = (int) (((double) Zoltan_Rand(NULL)/(double) ZOLTAN_RAND_MAX)*100000000);
```

Comment:
1. Use C++ - style casts, not C-style casts.
2. Don't use `NULL`; use `nullptr` instead.
3. Consider factoring out this unpleasant expression into a separate function.
4. Don't use magic numbers like `100000000`.  Give them names at least.

Reply:
This random number generation has been replaced with calls to std::srand and std::rand. I think all of these issues are resolved.